### ISSUE TITLE
Dynamic Task Mapping - Depth-first execution

### ISSUE THREAD

**Issue by Gollum999** [apache/airflow#25032]
### Description

I would like the ability to execute a group of dynamic tasks in a depth-first fashion.

This is not an original idea - the [original AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-42+Dynamic+Task+Mapping#AIP42DynamicTaskMapping-MapforTaskGroupsObjects) for Dynamic Task Mapping suggested using `TaskGroup`s for this feature, which seems like a great idea to me.  However I haven't seen any evidence indicating that this was still a planned feature, so I just wanted to make sure that this idea is not forgotten.

### Use case/motivation

Imagine a DAG like the following:
```
                     ┌──────────────┐ ┌────────────────┐ ┌───────────┐
                 ┌──►│extract_file_1│►│transform_file_1│►│load_file_1│
                 │   └──────────────┘ └────────────────┘ └───────────┘
                 │
                 │   ┌──────────────┐ ┌────────────────┐ ┌───────────┐
 ┌─────────────┐ ├──►│extract_file_2│►│transform_file_2│►│load_file_2│
 │get_file_list├─┤   └──────────────┘ └────────────────┘ └───────────┘
 └─────────────┘ │
                 │         ...               ...              ...
                 │
                 │   ┌──────────────┐ ┌────────────────┐ ┌───────────┐
                 └──►│extract_file_N│►│transform_file_n│►│load_file_N│
                     └──────────────┘ └────────────────┘ └───────────┘
```
Currently, this DAG would execute in breadth-first order, i.e. all mapped task instances of the same type must be complete before moving on to the next "phase".  This introduces unnecessary bottlenecks in the pipeline.  Each of these "branches" is independent, so there is no need to wait for `extract_file_N` before beginning `transform_file_1`, for example.

### Related issues

I couldn't find any.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by brandondtb** [comment]
I need this as well. I suspect it might be possible to accomplish this with SubDAGs, but haven't tested it yet,

**Comment by potiuk** [comment]
Yep. It's been initially excluded from the scope of dynamic task mapping implemented in 2.3.* but it would be good idea to have it at some point in time. If anyone wants to tackle it (though it is likely rather complex endevour) - feel free.

**Comment by uranusjr** [comment]
Note that even the initial scope only mentioned the mapped downstream would only depend on one mapped upstream. It does not guarantee depth-first semantic, but only the downstream _may_ be executed before other (unrelated) mapped upstreams.

Feel free to work on it, as mentioned above.

**Comment by AVMusorin** [comment]
It's very useful feature and I need it as well. @uranusjr May I try to implement it?

**Comment by potiuk** [comment]
You can try - but this is rather complex one and needs quite some testing I believe.

**Comment by wangyaopw** [comment]
+1

There is a highly desirable feature in the community. It would be awesome if Airflow can support this in the near future.

**Comment by potiuk** [comment]
It's already implemented in 2.5.0 - via TaskGroup Mapping.

**Comment by wangyaopw** [comment]
Hi, @potiuk 

I don't think it is possible to expand a taskgroup from the result of a task yet in 2.5.0.

For example

```
from airflow import DAG
from airflow.decorators import dag, task, task_group
from datetime import datetime

@task
def fetch_files():
    return ['file1', 'file2', 'file3']

@task_group
def etl_pipeline(next_file):
    @task
    def pre_process(file):
        return 'PreProcessed '+file

    @task
    def process(file):
        return 'Processed '+file

    @task
    def post_process(file):
        return 'PostProcessed '+file

    return post_process(process(pre_process(next_file)))


with DAG(dag_id="task_group_mapping_test",
         schedule="@daily",
         start_date=datetime(2022, 1, 1),
         catchup=False) as dag:
    # etl_pipeline.expand(next_file=['file1', 'file2', 'file3'])    <-- this runs fine

    # etl_pipeline.expand(next_file=fetch_files())     
    #  <-- this will cause scheduler error 
    # "airflow.models.expandinput.NotFullyPopulated: Failed to populate all mapping metadata; missing: 'next_file'
```

Is it possible to expand a taskgroup from task result in release 2.5?

I know expand a task from task result is possible.

**Comment by potiuk** [comment]
> Is it possible to expand a taskgroup from task result in release 2.5?

I believe it is @uranusjr - or I misunderstood it ?

**Comment by uranusjr** [comment]
Yes, but I would not recommend doing it since we still lack appropriate UI components to display those groups and tasks in them.

**Comment by talnagar** [comment]
@uranusjr do you know if there are open issues to address adding the UI components? It'll be a great feature to use.

**Comment by potiuk** [comment]
> Yes, but I would not recommend doing it since we still lack appropriate UI components to display those groups and tasks in them.

Ah. that would explain why we have not announced it yet.

**Comment by wangyaopw** [comment]
> Yes, but I would not recommend doing it since we still lack appropriate UI components to display those groups and tasks in them.

Hi @uranusjr, do you mind providing a code example? Using the example I provided? That would help a lot. Thanks!

**Comment by potiuk** [comment]
@wangyaopw As @uranusjr mentioned - it's better not to attempt it yet. There are bugs like just fixed https://github.com/apache/airflow/pull/28592 (which is likely the fix to your problem) which are preventing from using this feature **just yet**.

**Comment by valayDave** [comment]
Hey! Just curious to know if you have a timeline on when Depth first execution using TaskGroups will get integrated with Airflow.

**Comment by ntnhaatj** [comment]
Hi, I am not sure if it has been raised elsewhere yet, I would like to express here as it relates to the concept depth-first relationship I believe.

In general, I would like to have control over the upstream / downstream dependency in a depth-first manner when using dynamic mapped task group as the issue owner.
For instance, if I cleared the task `extract_file_2` along with its downstream, it should only clear its downstream only, which are `transform_file_2` and `load_file_2`. However, currently, it clears all downstream tasks, eg `transform_file_1`, `load_file_1`...

example code:

```python
from airflow.decorators import dag, task_group, task
from airflow.operators.empty import EmptyOperator
from pendulum import datetime

files = ["a", "b", "c"]

@dag(start_date=datetime(2022, 12, 1), schedule=None, catchup=False)
def task_group_mapping_example():
    @task_group(group_id="etl")
    def etl_pipeline(file):
        e = EmptyOperator(task_id="e")
        t = EmptyOperator(task_id="t")
        l = EmptyOperator(task_id="l")

        e >> t >> l

    etl = etl_pipeline.expand(file=files)
    etl

task_group_mapping_example()
```

it affects all mapped index downstream tasks if I clear individual mapped task (`etl.e[1]`)
<img width="877" alt="Screenshot 2023-08-19 at 18 46 42" src="https://github.com/apache/airflow/assets/35696768/dd110d7b-ecdf-4bcc-a466-a7d49b54ac87">

**Comment by uranusjr** [comment]
The depth-first execution model has been implemented. Opening a new issue is better if you would like to suggest improvements to it. Comments in old, closed issues are generally ignored.
